flink-runtime/src/test/java/org/apache/flink/runtime/resourcemanager/slotmanager/DeclarativeSlotManagerTest.java

```diff
@@ -0,0 +1,1092 @@
+			assertFalse(slotManager.reportSlotStatus(unknownInstanceID, unknownSlotReport));
+
+			assertThat(slotManager.getNumberRegisteredSlots(), is(0));
+		}
+	}
+
+	/**
+	 * Tests that slots are updated with respect to the latest incoming slot report. This means that
+	 * slots for which a report was received are updated accordingly.
+	 */
+	@Test
+	public void testUpdateSlotReport() throws Exception {
+		final TaskExecutorConnection taskManagerConnection = createTaskExecutorConnection();
+		final ResourceID resourceId = taskManagerConnection.getResourceID();
+
+		final SlotID slotId1 = new SlotID(resourceId, 0);
+		final SlotID slotId2 = new SlotID(resourceId, 1);
+
+		final SlotStatus slotStatus1 = createFreeSlotStatus(slotId1);
+		final SlotStatus slotStatus2 = createFreeSlotStatus(slotId2);
+
+		final SlotStatus newSlotStatus2 = createAllocatedSlotStatus(slotId2);
+		final JobID jobId = newSlotStatus2.getJobID();
+
+		final SlotReport slotReport1 = new SlotReport(Arrays.asList(slotStatus1, slotStatus2));
+		final SlotReport slotReport2 = new SlotReport(Arrays.asList(newSlotStatus2, slotStatus1));
+
+		final DefaultSlotTracker slotTracker = new DefaultSlotTracker();
+		try (DeclarativeSlotManager slotManager = createDeclarativeSlotManagerBuilder()
+			.setSlotTracker(slotTracker)
+			.buildAndStartWithDirectExec()) {
+
+			// check that we don't have any slots registered
+			assertEquals(0, slotManager.getNumberRegisteredSlots());
+
+			slotManager.registerTaskManager(taskManagerConnection, slotReport1);
+
+			DeclarativeTaskManagerSlot slot1 = slotTracker.getSlot(slotId1);
+			DeclarativeTaskManagerSlot slot2 = slotTracker.getSlot(slotId2);
+
+			assertEquals(2, slotManager.getNumberRegisteredSlots());
+
+			assertSame(SlotState.FREE, slot1.getState());
+			assertSame(SlotState.FREE, slot2.getState());
+
+			assertTrue(slotManager.reportSlotStatus(taskManagerConnection.getInstanceID(), slotReport2));
+
+			assertEquals(2, slotManager.getNumberRegisteredSlots());
+
+			assertNotNull(slotTracker.getSlot(slotId1));
+			assertNotNull(slotTracker.getSlot(slotId2));
+
+			// slotId2 should have been allocated for jobiD
+			assertEquals(jobId, slotTracker.getSlot(slotId2).getJobId());
+		}
+	}
+
+	/**
+	 * Tests that if a slot allocation times out we try to allocate another slot.
+	 */
+	@Test
+	public void testSlotAllocationTimeout() throws Exception {
+		final CompletableFuture<Void> secondSlotRequestFuture = new CompletableFuture<>();
+		final AtomicInteger slotRequestsCount = new AtomicInteger();
+		final TaskExecutorConnection taskManagerConnection = createTaskExecutorConnection(new TestingTaskExecutorGatewayBuilder()
+			.setRequestSlotFunction(ignored -> {
+				if (slotRequestsCount.getAndAdd(1) == 1) {
+					secondSlotRequestFuture.complete(null);
+				} else {
+					// mimic RPC timeout
+					return FutureUtils.completedExceptionally(new AskTimeoutException("timeout"));
+				}
+				return new CompletableFuture<>();
```

Comment:
A slightly easier way to express this logic would be to use a queue from which you pop and you populate the queue with the answers you want to have.